src/components/taskDetails/index.tsx

```diff
@@ -0,0 +1,198 @@
+import React, { ChangeEvent, FC } from 'react';
+import NavBar from '@/components/navBar/index';
+import TaskContainer from './TaskContainer';
+import Details from './Details';
+import classNames from './task-details.module.scss';
+import { toast, ToastTypes } from '@/helperFunctions/toast';
+import task from '@/interfaces/task.type';
+import { useState, useEffect, useContext } from 'react';
+import useFetch from '@/hooks/useFetch';
+import { isUserAuthorizedContext } from '@/context/isUserAuthorized';
+import fetch from '@/helperFunctions/fetch';
+
+type Props = {
+  url: string;
+  taskID: string;
+};
+
+const TaskDetails: FC<Props> = ({ url, taskID }) => {
+  const isAuthorized = useContext(isUserAuthorizedContext);
+  const [isEditing, setIsEditing] = useState<boolean>(false);
+  const [initialData, setInitialData] = useState<task>();
+  const [taskDetails, setTaskDetails] = useState<task>();
+  const [editedDetails, setEditedDetials] = useState({});
+  const { response, error, isLoading } = useFetch(url);
+  const { SUCCESS, ERROR } = ToastTypes;
+
+  useEffect(() => {
+    const fetchedData: task = { ...response.taskData };
+    setTaskDetails(fetchedData);
+    setInitialData(fetchedData);
+  }, [isLoading, response]);
+
+  function convertTimeStamp(timeStamp: string) {
+    const dateTime = new Date(parseInt(timeStamp));
+    return dateTime.toLocaleString();
+  }
+
+  function renderTextarea(name: string, value: string) {
+    return (
+      <textarea
+        className={classNames['textarea']}
+        name={name}
+        value={value}
+        onChange={(event) => handleChange(event)}
+      />
+    );
+  }
+
+  function handleChange(event: ChangeEvent) {
+    const formData: any = {
+      [(event.target as HTMLInputElement).name]: (
+        event.target as HTMLInputElement
+      ).value,
+    };
+    setEditedDetials({ ...editedDetails, ...formData });
+    setTaskDetails({ ...taskDetails, ...formData });
+  }
+
+  function onCancel() {
+    setIsEditing(false);
+    setEditedDetials({});
+    setTaskDetails(initialData);
+  }
+
+  async function onSave() {
+    setIsEditing(false);
+    try {
+      const response = fetch({
+        url: `${process.env.NEXT_PUBLIC_BASE_URL}/tasks/${taskID}`,
+        method: 'patch',
+        data: JSON.stringify(editedDetails),
+      });
+      const responseData = await response.requestPromise;
+      if (responseData.status === 204) {
+        setInitialData(taskDetails);
+        setEditedDetials({});
+        toast(SUCCESS, 'Successfully saved');
+      }
+    } catch (err) {
+      toast(ERROR, 'Could not save changes');
+      setTaskDetails(initialData);
```

Comment:
why are we setting `state` here again if it fails shouldn't it exit out of `editMode` or something

Reply:
When the super user clicks on save and by chance the request fails, we are making sure the UI is restored back to how it was before the changes were made.